### PR TITLE
Arbitrary orientations in parallel::distributed::Triangulation::add_periodicity

### PR DESCRIPTION
**Pull request by masterleinad**
This PR addresses the missing implementation for p4est_orientations other than 0.
Furthermore, the situation when rotated periodicity constraints require to constrain a DoF inversely to itself is considered.
The first test (periodicity_03) fails for p4est version 1.1 and before but is addressed in http://librelist.com/browser//p4est/2015/4/24/join-faces-for-faces-sharing-an-edge/ and the current development version (https://github.com/cburstedde/p4est/tree/develop).

### CODE REVIEW

File: include/deal.II/lac/constraint_matrix.h [hunk 1]
```diff
@@ -1089,6 +1089,18 @@ class ConstraintMatrix : public Subscriptor
    *
    * @ingroup Exceptions
    */
+  DeclException1 (ExcColumnNotStoredHere,
+                  size_type,
+                  << "The index set given to this constraint matrix indicates "
+                  << "constraints for degree of freedom " << arg1
+                  << " should not be stored by this object, but a constraint "
+                  << "uses this degree of freedom.");
```
Comment (bangerth) [inline include/deal.II/lac/constraint_matrix.h]: I don't think I understand this completely. Are you saying that a constraint *references* a DoF that is not part of the index set? (E.g., a constraint for a locally owned DoF references a locally *relevant* DoF, but the index set only includes the locally *active* DoFs?)

If so, could you rephrase the error message and maybe also include the index of the *constrained* DoF where you encounter the invalid index?

File: source/grid/grid_tools.cc [hunk 2]
```diff
@@ -2553,7 +2553,7 @@ namespace GridTools
     if (matrix.m() == spacedim)
       for (int i = 0; i < spacedim; ++i)
         for (int j = 0; j < spacedim; ++j)
-          distance(i) = matrix(i,j) * point1(j);
+          distance(i) += matrix(i,j) * point1(j);
```
Comment (bangerth) [inline source/grid/grid_tools.cc]: was this wrong before? how did nobody ever notice this...

**Comment by bangerth** [comment]
Nice. I'm not sure what to do about the testcase given that there is no p4est version we can currently use. @tamiko -- is there a way to condition the test on having at least p4est 1.2 (whenever that comes out)?

File: include/deal.II/lac/constraint_matrix.h (hunk 1 above)
Comment (masterleinad) [inline include/deal.II/lac/constraint_matrix.h]: Yes, you are right. This is for a constraint that references a DoF that is not part of the index set.
I rephrased the error message.

**Comment by bangerth** [comment]
@masterleinad -- let me know when you think this is ready. (Also, any insight on my comment on the change in grid_tools.cc?)

File: source/grid/grid_tools.cc (hunk 2 above)
Comment (masterleinad) [inline source/grid/grid_tools.cc]: periodicity_02.cc uses this line, but there it doesn't matter whether the matrix is {{0,1},{-1,0}} or {{0,1},{0,0}}. Apparently, make_periodicity_constraints with a rotation matrix has not been used for cases much different to the test case.

Comment (bangerth) [inline source/grid/grid_tools.cc]: Do I understand you right that the code before was simply wrong, but that nobody noticed because we happened to only run testcases where it didn't matter?

Comment (masterleinad) [inline source/grid/grid_tools.cc]: Yes, you're right. periodicity_03 now fails if this is not fixed.

**Comment by bangerth** [comment]
OK. Can you add an entry to changes.h?

File: include/deal.II/lac/constraint_matrix.h [hunk 3]
```diff
@@ -1361,6 +1374,8 @@ void
 ConstraintMatrix::add_line (const size_type line)
 {
   Assert (sorted==false, ExcMatrixIsClosed());
+  Assert (!local_lines.size() || local_lines.is_element(line),
```
Comment (tjhei) [inline include/deal.II/lac/constraint_matrix.h]: not really needed, because we will do this check in calculate_line_index() a few lines below.

File: include/deal.II/lac/constraint_matrix.h [hunk 4]
```diff
@@ -1406,6 +1421,8 @@ ConstraintMatrix::add_entry (const size_type line,
   // exists, since we don't want to enter it twice
   Assert (lines_cache[calculate_line_index(line)] != numbers::invalid_size_type,
           ExcInternalError());
+  Assert (!local_lines.size() || local_lines.is_element(column),
```
Comment (tjhei) [inline include/deal.II/lac/constraint_matrix.h]: Interesting. What is the reason for this? I so far thought that we won't need this. Do you have a test case where this is a problem? Do all tests still work?

Comment (bangerth) [inline include/deal.II/lac/constraint_matrix.h]: I think this is useful since we always had that locally active DoFs were constrained against locally relevant DoFs and we knew what the locally relevant DoFs were. But here the locally relevant DoFs may extend to the other side of the domain and one has to make sure that they are actually included in the locally relevant index set?

File: tests/mpi/periodicity_04.cc [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+  try
+    {
+      using namespace dealii;
+
+      Utilities::MPI::MPI_InitFinalize mpi_initialization(argc, argv, 1);
+      deallog.depth_console (0);
+
+      if (Utilities::MPI::this_mpi_process (MPI_COMM_WORLD)==0)
+        {
+          std::ofstream logfile("output");
```
Comment (tjhei) [inline tests/mpi/periodicity_04.cc]: I often prefer ```  MPILogInitAll log;``` which will capture output of all processes and makes ```main()``` much simpler. See ```mpi/tria_01.cc``` for an example.

**Comment by tjhei** [comment]
Looks very good. Thanks for working on this! :+1:

File: include/deal.II/lac/constraint_matrix.h (hunk 3 above)
Comment (masterleinad) [inline include/deal.II/lac/constraint_matrix.h]: OK. I remove this.

File: include/deal.II/lac/constraint_matrix.h (hunk 4 above)
Comment (masterleinad) [inline include/deal.II/lac/constraint_matrix.h]: @tjhei: No I don't have a test case for this. While working on dof_tools_constraints.cc this turned out to be useful.
@bangerth: I actually don't see how DoFs that extend to the other side of the domain can pose any problems. They are locally relevant if add_periodicity is used. 

The only problem that I can think of is if 
1.) the ConstraintMatrix is initialized with an IndexSet describing the locally active DoFs and 2.) constraints that reference locally relevant DoFs not in this IndexSet are used.
Basically, one is forced to initialize the ConstraintMatrix with the locally relevant IndexSet. Is this too restrictive?

File: tests/mpi/periodicity_04.cc (hunk 5 above)
Comment (masterleinad) [inline tests/mpi/periodicity_04.cc]: OK. I apply this at least for periodicity_04.cc

File: include/deal.II/lac/constraint_matrix.h (hunk 4 above)
Comment (bangerth) [inline include/deal.II/lac/constraint_matrix.h]: > @bangerth <https://github.com/bangerth>: I actually don't see how DoFs that
> extend to the other side of the domain can pose any problems. They are locally
> relevant if add_periodicity is used.

Ah, indeed. Good point. Thanks for clearing it up!

> The only problem that I can think of is if
> 1.) the ConstraintMatrix is initialized with an IndexSet describing the
> locally active DoFs and 2.) constraints that reference locally relevant DoFs
> not in this IndexSet are used.
> Basically, one is forced to initialize the ConstraintMatrix with the locally
> relevant IndexSet. Is this too restrictive?

No, I think that's exactly what needs to be done.

**Comment by tamiko** [comment]
@bangerth: Currently, we have no way of expressing version conditions for tests.

+1 for the patchset.

**Comment by tjhei** [comment]
Maybe we can disable the test for now (do not include .output files) until the next p4est release?

File: include/deal.II/lac/constraint_matrix.h (hunk 4 above)
Comment (tjhei) [inline include/deal.II/lac/constraint_matrix.h]: This check enforces that in a constraint ```x_i =x_j``` j also needs to be locally_relevant, right? My understanding is that this is not required in ```distribute_local_to_global``` because we can write anywhere. I am worried that this change might break some parallel codes (not from hanging node or periodicity constraints of course!).

Comment (masterleinad) [inline include/deal.II/lac/constraint_matrix.h]: You can always use a IndexSet other than the one obtained from locally relevant DoFs. In my opinion a constraint x_i=x_j is quite dangerous if j is not locally relevant. 
I understand your point, but is it really a problem to force the inclusion of such a non locally relevant DoF in the IndexSet of the ConstraintMatrix if this is indeed desired?

Comment (tjhei) [inline include/deal.II/lac/constraint_matrix.h]: I am only worried introducing this (especially right before the 8.3 release) without knowing if this breaks anything. Maybe I am just too cautious here. :-) 
@kronbichler what do you think, is this change safe?

**Comment by bangerth** [comment]
Yes, that may be an option -- just remove the .output file and open a github issue to remind ourselves that we need to add it at a later point.

@masterleinad -- can you do this? You also need to rebase the patch for merging.

File: include/deal.II/lac/constraint_matrix.h (hunk 4 above)
Comment (bangerth) [inline include/deal.II/lac/constraint_matrix.h]: I think the change is in general a good idea. I'd like to see the testsuite run on it. It may not be the right time for such a change a couple of days before a release, though :-)

@masterleinad : What do you think about removing this line from the patch for now and instead putting it into a separate pull request that we can postpone until after the next release?

**Comment by masterleinad** [comment]
The test periodicity_04.cc is referenced in Issue #1128.

File: include/deal.II/lac/constraint_matrix.h (hunk 4 above)
Comment (tjhei) [inline include/deal.II/lac/constraint_matrix.h]: yes, let us do that!

Comment (masterleinad) [inline include/deal.II/lac/constraint_matrix.h]: I am fine with postponing. See PR #1129

**Comment by tjhei** [comment]
Why did you remove _04 when you say in the PR message at the top that _03 is the one that fails? Is this a typo?

**Comment by tjhei** [comment]
Otherwise it looks good to be merged! Thanks again.

**Comment by masterleinad** [comment]
Your absolutely right. It should be periodicity_03.cc that is postponed and not periodicity_04.cc

**Comment by tjhei** [comment]
okay, is it good to go now?

**Comment by masterleinad** [comment]
Yes, from my point of view it can be merged.

**Comment by tjhei** [comment]
:+1:
